dist/privacy-sandbox.js

```diff
@@ -0,0 +1,78 @@
+//[privacy-sandbox]
+// Topics API Usage Reference: https://developers.google.com/privacy-sandbox/relevance/topics/demo#the-topics-api-demo
+// Header Usage Reference: https://developers.google.com/privacy-sandbox/relevance/topics/demo#the-topics-api-demo
+
+let requests = $WPT_BODIES;
+const firstPartyDomain = document.location.hostname;
+
+let result = {
+  topicsAvailable: document.featurePolicy.allowsFeature('browsing-topics'),
+  thirdPartiesUsingBrowsingTopics: {}
+}
+
+async function fetchAndCheckAttestation(url) {
+  const controller = new AbortController();
+  const { signal } = controller;
+  setTimeout(() => controller.abort(), 5000);
+
+  try {
+    const response = await fetch(url, { signal, mode: 'no-cors' });
+    return response ? true : false;
+  } catch (error) {
+    return false;
+  }
+}
+
+let seenThirdParties = [];
+(async () => {
+  for (const request of requests) {
+    const url = new URL(request.url);
+    const isScript = request.type === 'Script';
+    const isDocument = request.type === 'Document';
+
+    let thirdPartyDomain = '';
+    if (url.hostname !== firstPartyDomain) {
+      thirdPartyDomain = url.hostname;
+    }
+    if (thirdPartyDomain && seenThirdParties.includes(thirdPartyDomain)) {
+        continue;
+    }
+    seenThirdParties.push(thirdPartyDomain);
+
+    let jsUsage = false;
+    let headerUsage = false;
+    let receiverObserving = false;
+
+    // If the request is to fetch a javascript file or HTML document, perform string search in the returned file content for usage of Topics API
+    // document.browsingTopics() [JS] or fetch() request call includes: {browsingTopics: true} or XHR request call includes: {deprecatedBrowsingTopics: true} (to be deprecated)
+    if (isScript || isDocument) {
+      if (request.response_body && (request.response_body.includes('browsingTopics') || request.response_body.includes('deprecatedBrowsingTopics'))) {
+        jsUsage = true;
+      }
+    }
+
+    // Checking request header usage of Topics: header: 'Sec-Browsing-Topics: true'
+    if ('Sec-Browsing-Topics' in request.request_headers) {
+      headerUsage = true;
+      // Checking is sent Topics are observed by the receiver using response header 'Observe-Browsing-Topics'
+      // If value is ?1 then they are observed else they are not observed
+      if (request.response_headers['Observe-Browsing-Topics'] === '?1') {
+        receiverObserving = true;
+      }
+    }
+
+    if (thirdPartyDomain && (jsUsage || headerUsage)) {
```

Comment:
Are you suggesting this structure for each third-party or for the result object. I'm confused because currently, we are checking for attestation at third-party level and not at the test website level once. So, attestationPublished exists as flag for each third-party as follows:
```
attestationPublished: await fetchAndCheckAttestation(`${url.origin}/.well-known/privacy-sandbox-attestations.json`),
tldPlus1AttestationPublished: await fetchAndCheckAttestation(`${tldPlus1Origin}/.well-known/privacy-sandbox-attestations.json`)
```

tld+1 is separately done because ```bidder.criteo.com/.well-known/privacy-sandbox-attestations.json``` doesn't return attestation but ```criteo.com/.well-known/privacy-sandbox-attestations.json``` does. And almost all the times, criteo is embedded on a test website with origin as "gum.criteo.com" or "static.criteo.com" or "bidder.criteo.com" and never as "www.criteo.com" / "criteo.com". Some other origin could have implemented separate attestation for subdomains and the main domain. So, i handle both the cases.

Reply:
So there are 2 questions actually
1. attestation scope
2. extendable structure

Let's see:
1. This one is clear - attestation is done on [the site level](https://developers.google.com/privacy-sandbox/blog/announce-enrollment-privacy-sandbox), so the closest we are to it in the code is `getCanonicalDomain(hostname)` OR `tldPlus1Origin` (Are these duplicates?)

2. Extendable structure can be done either domain-centric (domains as object keys) or attribute-centric:
```js
"_privacy-sandbox": {
  'topicsAPI': {
    "topicsAvailable": true,
    "usingBrowsingTopics": ["criteo.com", "google.com"],
    "topicsAccessJs": ["criteo.com", "google.com"],
    "topicsAccessHeader": [],
    "observingTopics": ["criteo.com"],
  },
  'protected_audience': {
    ...
  },
  ...
  "attestationPublished": ["criteo.com", "google.com"]
}
```

Attribute-centric is more compact actually and works way better with native BigQuery functions, e.g.:
```sql
WITH parsed_data AS (
  SELECT
    page,
    JSON_EXTRACT_ARRAY(custom_metrics, '$._privacy-sandbox.topicsAPI.topicsAccessJs') AS topicsAccessJs,
    JSON_EXTRACT_ARRAY(custom_metrics, '$._privacy-sandbox.topicsAPI.topicsAccessHeader') AS topicsAccessHeader,
    JSON_EXTRACT_ARRAY(custom_metrics, '$._privacy-sandbox.attestationPublished') AS attestationPublished
  FROM
    dataset.table
),
attested_domains AS (
  SELECT
    page,
    domain
  FROM
    parsed_data,
    UNNEST(attestationPublished) AS domain
)

SELECT
  domain,
  COUNT(page) cnt
FROM
  attested_domains
GROUP BY domain
ORDER BY cnt DESC
```

What do you think?

Reply:
I agree attribution-specific extendable structure would be better. I have pushed it as the latest commit.